Add toolbar to YAML/template editors

TCWORLD opened the pull request:
Replaces the floating full-screen button over code editors with a handy configurable toolbar. This resolves an issue whereby the old button would float over the content of the editor and obscure it from view. 

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue or discussion
  in the additional information section.
-->

Please refer to issue #26092.

The full-screen button on code and template editor windows obscures the content of the editor if the content of that line is long enough to extends behind the button. 

Before the change, an editor (e.g. the `Dev Tools -> Template` window) with a long line of text would look like this:

<img width="466" height="227" alt="Image" src="https://github.com/user-attachments/assets/c7dbae15-b966-4e04-b35b-7c1fb856eed8" />

There is no way to see or interact with the code behind the full-screen button. 

To resolve this, based on feedback in this PR thread, the button has been moved onto a new editor toolbar to allow the content to always be visible. Along with the full-screen button, I've additionally added "Undo", "Redo" and "Copy" buttons to begin with as these seem to be useful features. The undo/redo control directly interfaces with the underlying CodeMirror state handling, so no need for any clever logic. The copy button uses the existing copy to clipboard mechanism.

<img width="454" height="258" alt="image" src="https://github.com/user-attachments/assets/93fd1989-cd32-4d1d-bdac-aa7450919278" /> <br>
<img width="643" height="192" alt="image" src="https://github.com/user-attachments/assets/11c3bd6b-60ac-443f-b94b-beaa64dbf905" />

As part of this, there is now a generic `ha-icon-button-toolbar` element which implements the logic of rendering the toolbar. This is not specifically tied to the code editor, and has a bit of configurability added to hopefully make it a useful element in itself. 


## Type of change
<!--
  What type of change does your PR introduce to the Home Assistant frontend?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR.
-->

Tested on a fresh development instance, no special configuration required.

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #26092

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

<!--
  Thank you for contributing <3
-->

[docs-repository]: https://github.com/home-assistant/home-assistant.io

TCWORLD committed:
Add paddingRight to cm-content
This prevents the fullscreen expansion button from obscuring the content of the editor box.

TCWORLD committed:
Add toolbar div above Code Editor

Add a new div above the cm-editor and move the fullscreen button into
it. This should probably be replaced with a proper HA element if one
exists.

TCWORLD committed:
Extend gutter border around cm-content top.

Now that we have a toolbar, it looks better if the border between the
content and line numbers extends up to be between the editor and toolbar
as well.

TCWORLD committed:
Move fullscreen button into ha-icon-button-group

Place in a button group, and make the editor toolbar a flex entity. This
allows the sizing and positioning of the buttons to be taken care of by
the button group rather than having to manually calculate positions.

TCWORLD committed:
Add copy to clipboard button to editor toolbar

Use the existing copyToClipboard utility to allow copying to contents of the editor view on button press.

TCWORLD committed:
Correct editor button icon colour.

Restore the CSS entry that sets the colour of the fullscreen button icon, and apply it to all buttons on the editor toolbar.

TCWORLD committed:
Add Undo Button

CodeMirror has this feature built in, so add a button for it. Should we also add redo?

TCWORLD committed:
Merge branch 'dev' into cm-content-padding

MindFreeze commented:
Good work @TCWORLD . We might have to refactor this whole component though to avoid all this manual DOM manipulation. It was a simple wrapper before but now with the extra buttons I think we need a proper `render` method.
@bramkragten what do you think?

TCWORLD committed:
Merge branch 'dev' into cm-content-padding

TCWORLD committed:
Refactor Code Editor Toolbar

Split the toolbar into its own LitElement class (ha-code-editor-toolbar). This makes ithe division between toolbar and code editor cleaner. Addition extra buttons is a simple case of putting extra entries in the toolbars items list.

TCWORLD committed:
Add `hasToolbar` Property to HA Code Editor

For cases where the toolbar may be unwanted, add the option to disable it. By default we have the new toolbar enabled.

Note that if the toolbar is disabled, there will be no fullscreen button.

karwosts commented:
Touching the code editor is tricky as it has some complex CSS requirements in many places. A lot of usecases like search broke the last time this was edited to add the fullscreen mode, we should be careful not to make it worse again.

Raw configuration editor needs to be height adjusted to avoid double scrollbar. 
<img width="222" height="309" alt="image" src="https://github.com/user-attachments/assets/64891a35-19c4-4d28-9d83-c2cc7d162a28" />

Fullpage automation code editor is overlapping copy buttons:

<img width="213" height="115" alt="image" src="https://github.com/user-attachments/assets/1ad8140b-478a-4af3-a54e-f79f65703101" />

TCWORLD committed:
Move Code Editor/Toolbar into Flex Div

This seems to stabilise a lot of the problems that were caused by adding the toolbar. The new div ensures that the toolbar is rendered at a fixed height, and then the remaining height is allocated to code-mirror.

TCWORLD commented:
@karwosts Would you be able to test again with the latest change?
I was getting different results to you (albeit just as glitchy). This change seems to have rectified a lot of the weirdness. Still not completely resolved on the automation editor.

TCWORLD committed:
Add CSS Variable to set Toolbar Height

Useful to have the parameter in case we want to resize the height of the toolbar.

karwosts commented:
> Double Toolbars gone from Raw Config Editor:

Also sorry to say but you removed the wrong scrollbar :joy:, and as a consequence the codemirror search is broken, (that's another one I fixed last month). 

The entire codemirror div must be on the screen, it can't be larger than the screen or else the native codemirror search doesn't work. So you have to force codemirror to be the screen size, and then the condent _inside_ the codemirror needs a scrollbar to scroll. 

This is why I was nervous about this 😓

TCWORLD committed:
Correct Toolbar DOM Placement

After discussions on PR #26580 its clear that the code mirror object must be the outer element as it doesn't play nice with other elements.

Instead of trying to put the toolbar above the editor, we should instead put it inside the editor view. By making it a child of the code mirror dom, and setting it to have an absolute position at the top, it resolves a lot of potential issues. The only problem then is that the toolbar overlaps the editor, fortunately this can easily be solved by padding the top of the cm-editor class with a space big enough for the toolbar - using padding rather than margins ensures the fragile size of the editor used for calculations is unaffected.

TCWORLD commented:
@karwosts Thanks for the pointers there. Helped me realise where I was going wrong.

I've pushed another change which seems to rectify all the problems I created, including with the automation editor (you're right, your other PR fixed the "production" off-screen weirdness I mentioned).

TCWORLD committed:
Add Button Lookup to Code Editor Toolbar

Helper function to find HaIconButton entities on the toolbar. Searches by element type, expected class, and optionally a secondary user class. The function returns undefined if no buttons are found.

TCWORLD committed:
Disable Undo/Redo Buttons if nothing to do

Add the redo button to go with undo (might as well have both if having one), and then make sure the buttons are only enabled if there is anything to undo or redo. This is purely done for aesthetic reasons as the buttons wouldn't have done anything anyway.

Similarly if the editor is read-only, don't allow the undo/redo buttons.

TCWORLD committed:
Always update redo/undo buttons on update

Otherwise if the same history state made no changes, the undo or redo button would be stuck left enabled.

TCWORLD committed:
Add ID to Editor Toolbar Buttons

This makes lookup of the buttons much simple using the also added findToolbarButtonById() method.

TCWORLD committed:
Rename HaCodeEditorToolbar to HaIconButtonToolbar

The toolbar itself is completely generic and not specific to the code editor. It basically is a wrapper to host a set of ha-icon-botton instances. Name it as such.

TCWORLD committed:
Merge branch 'dev' into cm-content-padding

TCWORLD committed:
Tidy findToolbarButtonById Calls

TCWORLD committed:
Fix Divider Option for ha-icon-butotn-toolbar

Was added as a placeholder but never implemented. If a divider element is added, it will add a vertical bar on the toolbar the same height as the icons with a little padding.

TCWORLD committed:
Merge branch 'dev' into cm-content-padding

TCWORLD commented:
Pretty happy with the state of this branch now, so I think it's ready for a review.

I can't check all the usages (e.g. ones on device setup pages where I don't have the appropriate device - e.g. Z-wave), but I've gone through all the editors I can find and the new toolbar appears to be functional and not mess up any of the CSS.


Current snagging points that I'd like feedback on:

- Copied to Clipboard toast appears behind the fullscreen editor on the Developer Tools pages (Template, Assist tabs). Don't know if I've caused this. If I have, could you advise on where this might be set? 
 
- The Automation fullscreen YAML editor has a "Copy to clipboard" button on it's bottom toolbar which is now duplicated by the copy button on the new editor toolbar. Which of the two should be removed?

- For the floating editor pane for Automations, when going fullscreen, the editor disappears behind the sidebar. Again I can't see how my changes could have caused this. Need to test the current dev branch when I get a chance to confirm it if its me causing it or whether its an existing bug.

    Collapsed:
    <img width="889" height="378" alt="image" src="https://github.com/user-attachments/assets/a5b3fd38-589f-418e-bebf-2bf0d2a01586" />

    Fullscreen:
    <img width="893" height="268" alt="image" src="https://github.com/user-attachments/assets/59a8b8a7-a8bc-4b0e-aa8b-7907dcbd9798" />

TCWORLD committed:
Reduce Code Editor z-index in Fullscreen

This means in the developer tools the fullscreen editor is no longer appearing on top of toast notifications.

karwosts commented:
Looks pretty nice, I can't find more problems. 👍 

For automation editor, I'd vote to remove copy to clipboard button in the footer. If we have the copy in the toolbar everywhere, might as well stay consistent. 

The only other place I think we have copy in the footer is in service action response, which has Copy Yaml and Copy JSON buttons, so that's perhaps not as simple to remove. But it doesn't create a problem for it to stay where it is.

TCWORLD committed:
Add Translation Entries for Code Editor Toolbar

Add localize entries for "Redo" (added in common where there was already  "Undo"), along with "Enter/Exit fullscreen" entries.

TCWORLD committed:
Use CSS to hide toolbar

Rather than overriding the toolbar style attribute, add a css entry to hide it when the host `:not(hasToolbar)`.

TCWORLD committed:
Apply localisation to toolbar

Some button labels were already using localisation, but incorrectly. Others were using hard coded english strings. Now they all use the entries added to the translation json.

TCWORLD committed:
Improve Editor Toolbar Creation Code

On creation of the code mirror object, we now only create the toolbar if required (hasToolbar = true). If not created at the start and the the flag later changes, the toolbar wiill be created at that point.

Furthermore, ensure that the `hasToolbar` class is properly set when the toolbar is created, otherwise the new CSS may hide it unintentionally.

TCWORLD committed:
Improve Code Editor Copy to Clipboard

Use `this.value` rather than trying to extract from the codemirror object. Skip copying if the editor is empty.

Furthermore, add a default toast message if localisation fails (e.g. missing hass element).

TCWORLD committed:
Remove Duplicate Clipboard Copy for YAML Editors

This is now on the code editor toolbar, so the extra button is no longer
required.

The logic for adding the copy button has been left in place in case of
usage where the code editor may be configured with no toolbar.

TCWORLD commented:
I've removed the duplicate copy to clipboard button as suggested by just removing the `copy-clipboard` attribute on the `ha-yaml-editor` where used. I wanted to try and limit changes, so I've left the logic that can add a copy button in the YAML editor class.

The button labels for the toolbar now also try to lookup localised strings (obvs these will need to be added for the various translations at some point) rather than being hard coded in English.

Finally a general tidy up of the code that adds the toolbar. It would sometimes disappear when moving between different editors (e.g. swapping between Templates and Events tabs in the dev tools). That's now been rectified.

TCWORLD committed:
Merge branch 'dev' into cm-content-padding

TCWORLD committed:
Improve Icon Button Toolbar Button Query

TCWORLD committed:
Merge branch 'dev' into cm-content-padding

TCWORLD committed:
Tidy fullscreen enter/exit

TCWORLD committed:
Fix Toolbar CSS Value

TCWORLD committed:
Merge branch 'dev' into cm-content-padding

TCWORLD committed:
Set fullscreen button label using property

TCWORLD committed:
Improve updating editor toolbar buttons

The mess of lookup up the icon buttons from the DOM to change their state turns out to be completely unnecessary. Instead we simply need to update the toolbar.items array whenever a button needs to change which will trigger the necessary re-rendering.

TCWORLD committed:
Merge branch 'dev' into cm-content-padding

TCWORLD committed:
Migrate ha-tooltip usage

Update to account for changes in #26540.

piitaya commented:
I tested it in card editor. The toolbar is not here when the field is empty.

piitaya commented on src/components/ha-code-editor.ts:
```diff
@@ -608,39 +732,33 @@ export class HaCodeEditor extends ReactiveElement {
     :host {
       position: relative;
       display: block;
+      --code-editor-toolbar-height: 36px;
```
```suggestion
      --code-editor-toolbar-height: 28px;
```

piitaya commented on src/components/ha-code-editor.ts:
```diff
@@ -666,15 +784,26 @@ export class HaCodeEditor extends ReactiveElement {
+      display: none !important;
+    }
+
+    .code-editor-toolbar {
+      --icon-button-toolbar-height: var(--code-editor-toolbar-height);
+      --icon-button-toolbar-color: var(
+        --code-editor-gutter-color,
+        var(--secondary-background-color, whitesmoke)
+      );
```
```suggestion
    .code-editor-toolbar {
      --icon-button-toolbar-height: var(--code-editor-toolbar-height);
      --icon-button-toolbar-color: var(
        --code-editor-gutter-color,
        var(--secondary-background-color, whitesmoke)
      );
      border-top-left-radius: var(--ha-border-radius-sm);
      border-top-right-radius: var(--ha-border-radius-sm);
```

piitaya commented on src/components/ha-icon-button-toolbar.ts:
```diff
@@ -0,0 +1,140 @@
+    .icon-toolbar-buttongroup {
+      background-color: transparent;
+      padding-right: 4px;
+      height: var(--icon-button-toolbar-height);
```
```suggestion
      height: var(--icon-button-toolbar-height);
      gap: 8px;
```

piitaya commented on src/components/ha-icon-button-toolbar.ts:
```diff
@@ -0,0 +1,140 @@
+        var(--icon-button-toolbar-height) - 4px
+      );
+      --icon-button-toolbar-icon: calc(
+        var(--icon-button-toolbar-height) - 14px
```
```suggestion
        var(--icon-button-toolbar-height) - 10px
```

piitaya reviewed (commented):
Added some css adjustments

TCWORLD commented on src/components/ha-icon-button-toolbar.ts:
```diff
@@ -0,0 +1,140 @@
+import "./ha-icon-button-group";
+import "./ha-tooltip";
+
+export type HaIconButtonToolbarItem =
+  | {
+      [key: string]: any;
+      divider?: false;
+      path: string;
+      label: string;
+      id?: string;
+      class?: string;
+      disabled?: boolean;
+      tooltip?: string;
+      action?: (e: Event) => any;
+    }
+  | {
+      [key: string]: any;
+      divider: true;
+      path?: never;
+      label?: never;
+      id?: never;
+      class?: never;
+      disabled?: never;
+      tooltip?: never;
+      action?: never;
+    };
```
This was originally based on the one from `ha-icon-overflow-menu.ts`, but I noticed that you couldn't make the items array containing a divider without giving a path/label for it (even though they were unused in that case). But I agree it was unnecessarily complex (could simply define path/label=""), so I'll change it back to what I had originally as a single interface.

TCWORLD commented:
> I tested it in card editor. The toolbar is not here when the field is empty.
> <img alt="CleanShot 2025-09-22 at 10 39 06" width="453" height="721" src="https://private-user-images.githubusercontent.com/5878303/492212485-d505d225-5589-431e-84cb-0ce4ca472627.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTg1NjM0MjEsIm5iZiI6MTc1ODU2MzEyMSwicGF0aCI6Ii81ODc4MzAzLzQ5MjIxMjQ4NS1kNTA1ZDIyNS01NTg5LTQzMWUtODRjYi0wY2U0Y2E0NzI2MjcucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDkyMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA5MjJUMTc0NTIxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Y2VmMjI1ZmY3NGM3ZGQ4NWFkYTU5YTAzYzdjMTExMmE2ZGQ3MzQxNDc3YjRhNzIwYmFmYjM0YjQ0Nzg4NzQxOCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.eA6VMSmabhr5CVcm2GaPGZtC-83Ja5Dnef8ctbMFotw">

Sorry I'm not familiar with that page, could you point me to which card editor you mean so that I can debug why?

TCWORLD committed:
Improvements to ha-icon-button-toolbar

TCWORLD committed:
Apply Toolbar Styling Suggestions

TCWORLD committed:
Coding style fixes

TCWORLD committed:
Fix Toolbar Missing for Empty Editors

Was caused by an error reading the length property of a potentially undefined value.

TCWORLD committed:
Merge branch 'dev' into cm-content-padding

TCWORLD committed:
Fix Code Editor Imports

TCWORLD committed:
Merge branch 'dev' into cm-content-padding

TCWORLD committed:
Apply Icon Button Toolbar change suggestions

Dividers can be added now by simply adding an empty string into the items list.

TCWORLD committed:
Apply CSS Change Suggestion

wendevlin reviewed (approved):
@TCWORLD thank you very much. That's a nice improvement.
